### PR TITLE
Optimize SnapshotTransaction.checkConstraints for read-only transactions

### PR DESCRIPTION
**Pull request by schlosna**
## General
**Before this PR**:
While reviewing some JFRs from high volume AtlasDB services (e.g. MP/GK), noticed significant allocations of `java.util.concurrent.ConcurrentHashMap$EntryIterator` pointing at `SnapshotTransaction.checkConstraints()`. We're doing a lot of unnecessary work during read-only transactions and transactions that hit tables without constraints.

image

**After this PR**:
==COMMIT_MSG==
Optimize SnapshotTransaction.checkConstraints for read-only transactions

When performing read-only transactions, we can avoid allocating and
accessing the constraints and writes tracking collections.
==COMMIT_MSG==

**Priority**: P1

**Concerns / possible downsides (what feedback would you like?)**:

**Is documentation needed?**: no

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 1]
```diff
@@ -1808,18 +1808,26 @@ public void commit(TransactionService transactionService) {
     }
 
     private void checkConstraints() {
-        List<String> violations = new ArrayList<>();
-        for (Map.Entry<TableReference, ConstraintCheckable> entry : constraintsByTableName.entrySet()) {
-            SortedMap<Cell, byte[]> sortedMap = writesByTable.get(entry.getKey());
-            if (sortedMap != null) {
-                violations.addAll(entry.getValue().findConstraintFailures(sortedMap, this, constraintCheckingMode));
-            }
+        if (constraintsByTableName.isEmpty() || writesByTable.isEmpty()) {
```
Comment (schlosna) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1811]: this will traverse both of these `ConcurrentHashMap`'s buckets to get a size() count, but will not allocate. In many cases `constraintsByTableName` will be empty when there are no constraints and short circuit, and in the case of read only transactions, `writesByTable` will be empty, so there is no work to check constraints.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 2]
```diff
@@ -1940,7 +1948,8 @@ protected void throwIfReadWriteConflictForSerializable(long commitTimestamp) {
     }
 
     private boolean hasWrites() {
-        return writesByTable.values().stream().anyMatch(writesForTable -> !writesForTable.isEmpty());
+        return !writesByTable.isEmpty()
```
Comment (schlosna) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1951]: This should avoid some of the expense of creating values & stream for read only transactions

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 1 above)
Comment (mdaudali) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1811]: Sadly, `constraintsByTableName will be empty when there are no constraints` is not strictly true - empty `constraintsByTable` implies no constraints, but no constraints does not imply an empty `constraintsByTable`. In fact, given the current implementation of SnapshotTransaction, there should be an entry for every unique table written to in the transaction, even if the constraint-providing method just returns the empty list. (This is because `SnapshotTransaction.useTable`, which is called when a table has been written to, creates a new entry in the `constraintsByTableName` map).

In summary, it would be sufficient to check just when `writesByTable` is empty, since entries in `writesByTable` implies at least one entry in `constraintsByTable`. This would only short-circuit for read-only transactions, and not when a table has no constraints.

Let me know if you'd like to take a shot at implementing no constraints => empty `constraintsByTable`, otherwise I can approve with the main read-only transaction improvement :)

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 3]
```diff
@@ -1808,18 +1808,26 @@ public void commit(TransactionService transactionService) {
...
+                .flatMap(e -> {
+                    NavigableMap<Cell, byte[]> writes = writesByTable.get(e.getKey());
+                    return writes == null
+                            ? Stream.empty()
+                            : e.getValue().findConstraintFailures(writes, this, constraintCheckingMode).stream();
+                })
+                .collect(toList());
+
         if (!violations.isEmpty()) {
+            AtlasDbConstraintException error = new AtlasDbConstraintException(violations);
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1826]: Positive: One nice side effect of this is that we should have the same stack traces if a client switches from exceptions <-> logging, which makes aggregating logs easier!

**Review by mdaudali — COMMENTED** [review]
lgtm, thanks for the contributions! Holding off approval to clarify the goal around skipping tables with no constraints :)

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 1 above)
Comment (schlosna) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java]: good call, when there are no constraints for a table, `findConstraintFailures` will return an empty list, so I've reworked the stream below to filter out when there are no constraints and avoid the flatMap stream allocation for those cases.

Comment (schlosna) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java]: actually going to just revert back to the for loop for simplicity and minimizing churn. The `List<String> violations = new ArrayList<>();` will be a cheap allocation and almost always empty assuming there are no constraint violations.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 4]
```diff
@@ -1808,18 +1808,25 @@ public void commit(TransactionService transactionService) {
...
+            return;
+        }
+
         List<String> violations = new ArrayList<>();
         for (Map.Entry<TableReference, ConstraintCheckable> entry : constraintsByTableName.entrySet()) {
-            SortedMap<Cell, byte[]> sortedMap = writesByTable.get(entry.getKey());
-            if (sortedMap != null) {
-                violations.addAll(entry.getValue().findConstraintFailures(sortedMap, this, constraintCheckingMode));
+            Map<Cell, byte[]> writes = writesByTable.get(entry.getKey());
+            if (writes != null) {
```
Comment (schlosna) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1819]:
```suggestion
            if (writes != null && !writes.isEmpty()) {
```

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 5]
```diff
@@ -1808,18 +1808,25 @@ public void commit(TransactionService transactionService) {
...
+                violations.addAll(entry.getValue().findConstraintFailures(writes, this, constraintCheckingMode));
```
Comment (schlosna) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1820]:
```suggestion
                List<String> constraintFailures = entry.getValue().findConstraintFailures(writes, this, constraintCheckingMode);
                if (!constraintFailures.isEmpty()) {
                    violations.addAll(constraintFailures);
                }
```

**Comment by schlosna** [comment]
Thanks for the review @mdaudali ! I pushed a couple commits to address your comments and updated a couple other read paths to avoid work when there are no writes.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 1 above)
Comment (mdaudali) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java]: Thanks! One other thing I'm going to note for posterity is that if any ConstraintCheckable had side effects _and_ relied  on them happening on read transactions (meaning that incorrectly marked table as used), then this does break that. However, 1) I couldn't find any internal uses of that and 2) should not be marking a table as used when there are only reads, therefore I'm okay with this.

**Review by mdaudali — APPROVED** [review]
Thanks for the contribution and taking this over the line! lgtm :)
